src/main/kotlin/com/github/biomejs/intellijbiome/services/BiomeServerService.kt

```diff
@@ -124,18 +125,39 @@ class BiomeServerService(private val project: Project) {
 
             val formattingResults = server.sendRequest { it.textDocumentService.formatting(formattingParams) }
 
-            formattingResults?.forEach {
-                val lineSeparator = StringUtil.detectSeparators(it.newText)
-                val normalizedText = StringUtil.convertLineSeparators(it.newText)
+            formattingActions.add(Runnable {
+                var lineSeparator: LineSeparator? = null
+
+                formattingResults?.forEach {
+                    val range = getRangeInDocument(document, it.range) ?: return@forEach
+
+                    if (StringUtil.isEmpty(it.newText)) {
+                        document.deleteString(range.startOffset, range.endOffset)
+                    } else {
+                        val normalizedText = StringUtil.convertLineSeparators(it.newText)
+
+                        if (range.endOffset >= 0) {
+                            if (range.length <= 0) {
+                                document.insertString(range.startOffset, normalizedText)
+                            } else {
+                                document.replaceString(range.startOffset, range.endOffset, normalizedText)
+                            }
+                        } else if (range.startOffset > 0) {
+                            document.insertString(range.startOffset, normalizedText)
+                        } else if (!StringUtil.equals(document.charsSequence, normalizedText)) {
+                            document.setText(normalizedText)
+                        }
+                    }
 
-                formattingActions.add(Runnable {
-                    if (!StringUtil.equals(document.charsSequence, normalizedText)) {
-                        document.setText(normalizedText)
+                    StringUtil.detectSeparators(it.newText)?.apply {
```

Comment:
The new text may or may not contain line separators. We will set the detected line separator only if any of the text edits contains a text separator.